flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.flink.table.catalog.ResolvedSchema;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * KNN is to classify unlabeled observations by assigning them to the class of the most similar
+ * labeled examples.
+ */
+public class Knn implements Estimator<Knn, KnnModel>, KnnParams<Knn> {
+
+    private static final long serialVersionUID = 5292477422193301398L;
+    private static final int ROW_SIZE = 2;
+    private static final int FASTDISTANCE_TYPE_INDEX = 0;
+    private static final int DATA_INDEX = 1;
+
+    protected Map<Param<?>, Object> params = new HashMap<>();
+
+    /** constructor. */
+    public Knn() {
+        ParamUtils.initializeMapWithDefaultValues(params, this);
+    }
+
+    /**
+     * constructor.
+     *
+     * @param params parameters for algorithm.
+     */
+    public Knn(Map<Param<?>, Object> params) {
+        this.params = params;
+    }
+
+    /**
+     * @param inputs a list of tables
+     * @return knn classification model.
+     */
+    @Override
+    public KnnModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        ResolvedSchema schema = inputs[0].getResolvedSchema();
+        String[] colNames = schema.getColumnNames().toArray(new String[0]);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Row> input = tEnv.toDataStream(inputs[0]);
+        String[] targetCols = getFeatureCols();
+        final int[] featureIndices;
+        if (targetCols == null) {
+            featureIndices = new int[colNames.length];
+            for (int i = 0; i < colNames.length; i++) {
+                featureIndices[i] = i;
+            }
+        } else {
+            featureIndices = new int[targetCols.length];
+            for (int i = 0; i < featureIndices.length; i++) {
+                featureIndices[i] = findColIndex(colNames, targetCols[i]);
+            }
+        }
+        String labelCol = getLabelCol();
+        final int labelIdx = findColIndex(colNames, labelCol);
+        final int vecIdx =
+                getVectorCol() != null
+                        ? findColIndex(
+                                inputs[0]
+                                        .getResolvedSchema()
+                                        .getColumnNames()
+                                        .toArray(new String[0]),
+                                getVectorCol())
+                        : -1;
+
+        DataStream<Row> trainData =
+                input.map(
+                        (MapFunction<Row, Row>)
+                                value -> {
+                                    Object label = value.getField(labelIdx);
```

Comment:
Can we use `value.getField(getLabelCol())` here?

Reply:
OK, using this pattern may be better. I will modify it later.

Reply:
done